osf/utils/registrations.py

```diff
@@ -0,0 +1,174 @@
+        key = keys.pop(0)
+        # Returns the value associated with the given key
+        value = get_value_or_extra(nested_response, block_type, key, keys)
+        return get_nested_answer(value, block_type, keys)
+    else:
+        # Once we've drilled down through the entire dictionary, our nested_response
+        # should be an array or a string
+        return nested_response
+
+# For expanding registration_responses
+def set_nested_values(nested_dictionary, keys, value):
+    """
+    Drills down through the nested dictionary, accessing each key in the array,
+    and sets the last key equal to the  passed in value, if this key doesn't already exist.
+
+    Assumes all keys are present, except for potentially the final key.
+
+    :param nested_dictionary, dictionary
+    :param keys, array
+    :param value, object, array, or string
+    """
+    for key in keys[:-1]:
+        nested_dictionary = nested_dictionary.get(key, None)
+
+    final_key = keys[-1]
+    if not nested_dictionary.get(final_key):
+        nested_dictionary[final_key] = value
+
+# For expanding registration_responses
+def build_answer_block(block_type, value):
+    extra = []
+    if block_type == 'file-input':
+        extra = value
+        value = ''
+        for file in extra:
+            if file.get('file_name'):
+                file['data'] = {
+                    'name': file.get('file_name')  # What legacy FE needs for rendering
+                }
+    return {
+        'comments': [],
+        'value': value,
+        'extra': extra
+    }
+
+# For expanding registration_responses
+def build_registration_metadata_dict(keys, current_index=0, metadata={}, value={}):
+    """
+    Function will recursively loop through each key in the list, checking if it exists in metadata, if not,
+    adding another nested level in the dictionary.
+
+    For example, calling build_registration_metadata_dict(
+        ["recommended-analysis", "value", "specify", "value", "question11c"], value='hello'), yields,
+
+    {
+        'recommended-analysis': {
+            'value': {
+                'specify': {
+                    'value': {
+                        'question11c': 'hello'
+                    }
+                }
+            }
+        }
+    }
+    :param keys array, of nested question_ids: ["recommended-analysis", "value", "specify", "value", "question11c"]
+    :param current_index, call initially with 0
+    :param metadata - registration_metadata
+    :param value, provide most deeply nested value.
+    :returns partial registration_metadata
+    """
+    if current_index == len(keys):
+        # We've iterated through all the keys, so we exit.
+        return metadata
+    else:
+        # All keys from left to right including the current key
+        current_chain = keys[0:current_index + 1]
+        # If we're on the final key, use the passed in value
+        val = value if current_index == len(keys) - 1 else {}
```

Comment:
The final-key case is not already handled above, because `set_nested_values` is called multiple times, each time with an increasingly longer list.  So while it might be the final key of the passed in list for that loop, it's not going to be the final key of the original list until the last call.   Intermediate, shorter, lists need to be set to empty dictionaries if the value doesn't exist. 

I'll give you an example. If keys are ['q7', 'value', 'question'], `set_nested_values` will be hit three times:
1) ['q7'], final key should be set to empty dict
2) ['q7', 'value'], final key should be set to empty dict
3) ['q7', 'value', 'question'], final call, so final key can be set to actual value

```
    'q7': {
        'value': {
            'question': {
                'comments': [],
                'value': 'data collection procedures',
                'extra': []
            }
      }
}
```

Reply:
Oh whoops, sorry! Even though I pointed out it was hard to read, I still mis-parsed the line -- thought the condition was `current_index == len(keys)`, missed the `- 1`. Ternary operators get confusing fast as their operands get more complex than a single name.

Reply:
no worries, I agree it's hard to read, trying to build this more expanded dictionary is one of the uglier things I've written, though part of the ugliness is the task itself, the fact that we have to expand in the first place.

I'm not going to have time to change my implementation though, I'm more focused on trying to test this and make sure the expanded version has everything we need to register, for example, sorry.